### PR TITLE
 [#187364729] User Login

### PR DESCRIPTION
**Pull request by Furaha-Justine**
**PR Description**

**Summary**
This pull request introduces a user login feature to enable users to authenticate and access their accounts using email and password within the application.

**Changes Made**

- Added a new LoginController class responsible for handling user login operations.
- Implemented user authentication logic, including password validation and JWT token generation.
- Integrated bcrypt for secure password hashing to enhance user data security.
- Included validation of the login request payload to ensure data integrity and prevent malformed requests.

**Reason for Changes**

- Enhanced User Experience: Enabling user login functionality enhances the overall user experience by providing personalized access to their accounts and data.
- Security Measures: The implementation of bcrypt for password hashing and JWT for token-based authentication enhances the security of user credentials and sessions.
- Data Integrity: Validating the login request payload ensures that only properly formatted requests are processed, reducing the risk of potential security vulnerabilities.

**Testing**
[https://www.loom.com/share/bb6c2c6c97ee49c883d31e68797ca3be?sid=1d14a929-94a5-4146-85bd-7a50274503dd](url)

**Additional Notes**

- This pull request includes thorough testing of the user login functionality to ensure reliability and correctness.
- Feedback on the implementation, code structure, and security considerations is highly appreciated to further improve the feature.

**Pivotal Tracker Story**
#187364729

### CODE REVIEW

**Review by mpanoathos — APPROVED** [review]

**Review by EricNiyo73 — APPROVED** [review]

File: .nycrc.json [hunk 1]
```diff
@@ -5,6 +5,7 @@
     "node_modules/**",
     "coverage/**",
     "src/config/database.ts",
+    "src/migrations/*.js",
```
Comment (munezerobagira) [inline .nycrc.json]: why did change this file?

File: src/database/migrations/20240427140502-create-user.js
Comment (munezerobagira) [inline src/database/migrations/20240427140502-create-user.js:1]: why did you move this file?

Comment (munezerobagira) [reply src/database/migrations/20240427140502-create-user.js:1]: normally migration are not supposed to be edited

**Review by munezerobagira — CHANGES_REQUESTED** [review]

File: .nycrc.json (hunk 1 above)
Comment (Furaha-Justine) [reply .nycrc.json]: When I was running tests it showed me coverage of migrations and I think it was not supposed to be there.

File: src/database/migrations/20240427140502-create-user.js
Comment (Furaha-Justine) [reply src/database/migrations/20240427140502-create-user.js:1]: Where? I didn't move it.

File: .nycrc.json (hunk 1 above)
Comment (Furaha-Justine) [reply .nycrc.json]: That is why I added it in exclude.

Comment (munezerobagira) [reply .nycrc.json]: that means you moved migrations files

Comment (munezerobagira) [inline .nycrc.json]:
```suggestion
```

**Review by munezerobagira — CHANGES_REQUESTED** [review]

File: src/database/migrations/20240427140502-create-user.js
Comment (munezerobagira) [reply src/database/migrations/20240427140502-create-user.js:1]: it   seems that you deleted migration and created new one, or manually renamed the file

Comment (Furaha-Justine) [reply src/database/migrations/20240427140502-create-user.js:1]: Oooh because I created it before you merge user registration but they have same codes cause I used it to avoid confusion.

**Review by munezerobagira — APPROVED** [review]

File: src/controllers/LoginController.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+    const isPasswordValid = await bcrypt.compare(
+      password,
+      existingUser.password,
+    )
+    if (!isPasswordValid) {
+      return res.status(401).json({ message: 'Invalid email or password' })
+    }
+
+    const token = jwt.sign({ id: existingUser.id }, process.env.JWT_SECRET)
```
Comment (munezerobagira) [inline src/controllers/LoginController.ts]:
```suggestion
    const token = jwt.sign({ id: existingUser.id }, process.env.JWT_SECRET)
```

if possible you can extract this function because it might be used at multiple place

**Review by munezerobagira — CHANGES_REQUESTED** [review]

**Review by munezerobagira — APPROVED** [review]
